src/dag-node-utils.js

```diff
@@ -0,0 +1,210 @@
+    return link
+  })
+
+  // Sort the links (in-place)
+  sort(links, linkSort)
+
+  serialize({
+    data: data,
+    links: links
+  }, (err, serialized) => {
+    if (err) {
+      callback(err)
+    }
+    hash(hashAlg, serialized, (err, multihash) => {
+      if (err) {
+        callback(err)
+      }
+      const dagNode = new DAGNode(data, links, serialized, multihash)
+      callback(null, dagNode)
+    })
+  })
+}
+
+function addLink (dagNode, nameOrLink, nodeOrMultihash, callback) {
+  const links = _cloneLinks(dagNode)
+  const data = _cloneData(dagNode)
+  let newLink = null
+
+  if ((nameOrLink.constructor &&
+       nameOrLink.constructor.name === 'DAGLink')) {
+    // It's a link
+    newLink = nameOrLink
+  } else if (typeof nameOrLink === 'string') {
+    // It's a name
+    if ((nodeOrMultihash.constructor &&
+       nodeOrMultihash.constructor.name === 'DAGNode')) {
+      // It's a node
+      // haadcode: not sure what ^ means, so the line below might not be correct
+      newLink = toDAGLink(nodeOrMultihash)
+    } else {
+      // It's a multihash
+      // haadcode: not sure what ^ means, so the line below might not be correct
+      newLink = new DAGLink(null, dagNode.size, nodeOrMultihash)
+    }
+  }
+
+  if (newLink) {
+    links.push(newLink)
+    sort(links, linkSort)
+  } else {
+    throw new Error('Link given as the argument is invalid')
+  }
+
+  create(data, links, callback)
+}
+
+function removeLink (dagNode, nameOrMultihash, callback) {
+  let data = _cloneData(dagNode)
+  let links = _cloneLinks(dagNode)
+
+  if (typeof nameOrMultihash === 'string') {
+    links = dagNode.links.filter((link) => link.name !== nameOrMultihash)
+  } else if (Buffer.isBuffer(nameOrMultihash)) {
+    links = dagNode.links.filter((link) => !link.hash.equals(nameOrMultihash))
+  } else {
+    throw new Error('second arg needs to be a name or multihash')
+  }
+
+  create(data, links, callback)
+}
+
+/*
+ * toDAGLink converts a DAGNode to a DAGLink
+ */
+// was: makeLink(node, callback)
+function toDAGLink (dagNode) {
+  return new DAGLink(null, dagNode.size, dagNode.multihash)
+}
+
+function clone (dagNode, callback) {
+  const data = _cloneData(dagNode)
+  const links = _cloneLinks(dagNode)
+  create(data, links, callback)
+}
+
+function serialize (node, callback) {
+  let serialized
+
+  try {
+    const pb = toProtoBuf(node)
+    serialized = proto.PBNode.encode(pb)
+  } catch (err) {
+    return callback(err)
+  }
+
+  callback(null, serialized)
+}
+
+function deserialize (data, callback) {
+  const pbn = proto.PBNode.decode(data)
+
+  const links = pbn.Links.map((link) => {
+    return new DAGLink(link.Name, link.Tsize, link.Hash)
+  })
+
+  sort(links, linkSort)
+
+  const buf = pbn.Data || new Buffer(0)
+
+  create(buf, links, callback)
+}
+
+function toProtoBuf (node) {
+  const pbn = {}
+
+  if (node.data && node.data.length > 0) {
+    pbn.Data = node.data
+  } else {
+    pbn.Data = null // new Buffer(0)
```

Comment:
please don't change this, it needs to be a Buffer of size 0, we had a problem in the past because the serialized versions were different when data was empty (go vs js)